### ISSUE TITLE
question_mark suggestion changed code semantics

### ISSUE THREAD

**Issue by divagant-martian** [rust-lang/rust-clippy#9518]
### Summary

If let expression that produces a result is suggested to change to `?`, changing the semantics of the code

### Lint Name

question_mark

### Reproducer

**I tried [this](https://play.integer32.com/?version=stable&mode=debug&edition=2021&gist=9959851def574b60f276213e7dc01c6e) code:**

```rust
fn do_stuff() -> Result<(), ()> {
    let result = if let Err(e) = op1() { Err(e) } else { Ok(()) };

    do_something_with_result(&result);

    result
}

fn op1() -> Result<(), ()> {
    Err(())
}

fn do_something_with_result(arg: &Result<(), ()>) {
    println!("doing something with the result {arg:?}");
}

fn main() {
    do_stuff().unwrap();
}

}

```

**I saw this happen:**

```
warning: this block may be rewritten with the `?` operator
 --> src/main.rs:2:18
  |
2 |     let result = if let Err(e) = op1() { Err(e) } else { Ok(()) };
  |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `op1()?`
  |
  = note: `#[warn(clippy::question_mark)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#question_mark
```

I **expected to see this happen:**
There should be no warning. Changing the code for what clippy suggests would prevent calling `do_something_with_result` when the result is an error

### Version

```text
rustc 1.64.0 (a55dd71d5 2022-09-19)
binary: rustc
commit-hash: a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52
commit-date: 2022-09-19
host: x86_64-unknown-linux-gnu
release: 1.64.0
LLVM version: 14.0.6
```

**Comment by ebobrow** [comment]
If I'm not mistaken, it looks like there are two issues here. Clippy's suggested code wouldn't even compile because the question mark returns the inner value of the `Result` while the original code returns a new `Result` (I think related to https://github.com/rust-lang/rust-clippy/issues/8281#issuecomment-1212597668). Do you have a reproduction where Clippy's suggestion is valid code, but we still get this semantics error? Or is this issue fixed by fixing the invalid suggestion?
